src/inmanta/server/agentmanager.py

```diff
@@ -1295,6 +1267,75 @@ async def _fork_inmanta(
             if errhandle is not None:
                 errhandle.close()
 
+    async def _wait_for_agents(
+        self, env: data.Environment, agents: Set[str], *, connection: Optional[asyncpg.connection.Connection] = None
+    ) -> None:
+        """
+        Wait until all requested autostarted agent instances are active, e.g. after starting a new agent process.
+
+        Must be called under the agent lock.
+
+        :param env: The environment for which to wait for agents.
+        :param agents: Autostarted agent endpoints to wait for.
+
+        :raises TimeoutError: When not all agent instances are active and no new agent instance became active in the last
+            5 seconds.
+        """
+        agent_statuses: dict[str, Optional[AgentStatus]] = await data.Agent.get_statuses(env.id, agents, connection=connection)
+        # Only wait for agents that are not paused
+        expected_agents_in_up_state: Set[str] = {
+            agent_name
+            for agent_name, status in agent_statuses.items()
+            if status is not None and status is not AgentStatus.paused
+        }
+
+        assert env.id in self._agent_procs
+        proc = self._agent_procs[env.id]
+
+        actual_agents_in_up_state: set[str] = set()
+        started = int(time.time())
+        last_new_agent_seen = started
+        last_log = started
+
+        while len(expected_agents_in_up_state) != len(actual_agents_in_up_state):
+            await asyncio.sleep(0.1)
+            now = int(time.time())
+            if now - last_new_agent_seen > AUTO_STARTED_AGENT_WAIT:
+                LOGGER.warning(
+                    "Timeout: agent with PID %s took too long to start: still waiting for agent instances %s",
+                    proc.pid,
+                    ",".join(sorted(expected_agents_in_up_state - actual_agents_in_up_state)),
+                )
+                raise asyncio.TimeoutError()
+            if now - last_log > AUTO_STARTED_AGENT_WAIT_LOG_INTERVAL:
+                last_log = now
+                LOGGER.debug(
+                    "Waiting for agent with PID %s, waited %d seconds, %d/%d instances up",
+                    proc.pid,
+                    now - started,
+                    len(actual_agents_in_up_state),
+                    len(expected_agents_in_up_state),
+                )
+            new_actual_agents_in_up_state = {
+                agent_name
+                for agent_name in expected_agents_in_up_state
+                if (
+                    (session := self._agent_manager.tid_endpoint_to_session.get((env.id, agent_name), None)) is not None
+                    # make sure to check for expiry because sessions are unregistered from the agent manager asyncronously
+                    and not session.expired
```

Comment:
I added this condition: a session might still be in the `tid_endpoint_to_session` mapping while it has already been expired (if the agent manager hasn't processed the expiry event yet).